options-design/mongo/clientopts/clientopts.go

```diff
@@ -0,0 +1,90 @@
+package clientopts
+
+import (
+	"context"
+	"net"
+	"time"
+)
+
+type ContextDialer interface {
+	DialContext(ctx context.Context, network, address string) (net.Conn, error)
+}
+
+type Option interface {
+	option()
+}
+
+type OptionFunc func(*Client) error
+
+type Client struct{}
+
+type Bundle struct{}
+
+func BundleOptions(...Option) Bundle { return Bundle{} }
+
+func (b *Bundle) Appname(string) *Bundle                                  { return nil }
+func (b *Bundle) AuthMechanism(string) *Bundle                            { return nil }
+func (b *Bundle) AuthMechaismProperties(map[string]string) *Bundle        { return nil }
+func (b *Bundle) AuthSource(s string) *Bundle                             { return nil }
+func (b *Bundle) ConnectTimeout(d time.Duration) *Bundle                  { return nil }
+func (b *Bundle) Dialer(d ContextDialer) *Bundle                          { return nil }
+func (b *Bundle) HeartbeatInterval(d time.Duration) *Bundle               { return nil }
+func (b *Bundle) Hosts(s []string) *Bundle                                { return nil }
+func (b *Bundle) Journal(v bool) *Bundle                                  { return nil }
+func (b *Bundle) LocalThreshold(d time.Duration) *Bundle                  { return nil }
+func (b *Bundle) MaxConnIdleTime(d time.Duration) *Bundle                 { return nil }
+func (b *Bundle) MaxConnsPerHost(u uint16) *Bundle                        { return nil }
+func (b *Bundle) MaxIdleConnsPerHost(u uint16) *Bundle                    { return nil }
+func (b *Bundle) Password(s string) *Bundle                               { return nil }
+func (b *Bundle) ReadConcernLevel(s string) *Bundle                       { return nil }
+func (b *Bundle) ReadPreference(s string) *Bundle                         { return nil }
```

Comment:
Can they just provide a ReadPreference object?

Reply:
Read preference can't just be a string, unless it's extended JSON.  It has to include not only the mode but also replica set tags.

Reply:
This is copied from our current client options. Right now users can specify a read preference string and tag sets separately. These are two different fields on the connection string as well, so with the current way it works a user could set tags in the uri and then also provide an option that overrides the read preference and have the two combined. If we only take a `*readpref.ReadPref` we won't be able to do that.